pkg/river/river.go

```diff
@@ -0,0 +1,322 @@
+	f := builder.NewFile()
+	f.Body().AppendFrom(v)
+
+	_, err := f.WriteTo(enc.w)
+	return err
+}
+
+// EncodeValue converts the value pointed to by v into a River value and writes
+// the result to the Decoder's output stream.
+//
+// See the documentation for MarshalValue for details about the conversion of
+// Go values into River values.
+func (enc *Encoder) EncodeValue(v interface{}) error {
+	expr := builder.NewExpr()
+	expr.SetValue(v)
+
+	_, err := expr.WriteTo(enc.w)
+	return err
+}
+
+// Unmarshal converts the River configuration file specified by in and stores
+// it in the struct value pointed to by v. If v is nil or not a pointer,
+// Unmarshal panics. The configuration specified by in may use expressions to
+// compute values while unmarshaling. Refer to the River language documentation
+// for the list of valid formatting and expression rules.
+//
+// Unmarshal uses the inverse of the encoding rules that Marshal uses,
+// allocating maps, slices, and pointers as necessary.
+//
+// To unmarshal a River body into a struct, Unmarshal matches incoming
+// attributes and blocks to the river struct tags specified by v. Incoming
+// attribute and blocks which do not match to a river struct tag cause a
+// decoding error. Additionally, any attribute or block marked as required by
+// the river struct tag that are not present in the source text will generate a
+// decoding error.
+//
+// To unmarshal a list of River blocks into a slice, Unmarshal resets the slice
+// length to zero and then appends each element to the slice.
+//
+// To unmarshal a list of River blocks into a Go array, Unmarshal decodes each
+// block into the corresponding Go array element. If the number of River blocks
+// does not match the length of the Go array, a decoding error is returned.
+//
+// Unmarshal follows the rules specified by UnmarshalValue when unmarshaling
+// the value of an attribute.
+func Unmarshal(in []byte, v interface{}) error {
+	dec := NewDecoder(bytes.NewReader(in))
+	return dec.Decode(v)
+}
+
+// UnmarshalValue converts the River configuration file specified by in and
+// stores it in the value pointed to by v. If v is nil or not a pointer,
+// UnmarshalValue panics. The configuration specified by in may use expressions
+// to compute values while unmarshaling. Refer to the River language
+// documentation for the list of valid formatting and expression rules.
+//
+// Unmarshal uses the inverse of the encoding rules that MarshalValue uses,
+// allocating maps, slices, and pointers as necessary, with the following
+// additional rules:
+//
+// After converting a River value into its Go value counterpart, the Go value
+// may be converted into a capsule if the capsule type implements
+// ConvertibleIntoCapsule.
+//
+// To unmarshal a River object into a struct, UnmarshalValue matches incoming
+// object fields to the river struct tags specified by v. Incoming object
+// fields which do not match to a river struct tag cause a decoding error.
```

Comment:
That means that superfluous fields will cause a decoding error, and for the unmarshaling to work correctly the incoming fields must match 1:1 the exported fields that have a `river` tag, right?

Reply:
Yeah, if you want to support arbitrary field names you need to unmarshal into a map instead. 

Should I make that more clear in the comments?

Reply:
No, just wanted to verify. Our unmarshaller will be a little stricter than say json.Unmarshal which IIRC allows duplicate or missing fields, but I don't think that's a bad thing.